Userland/Libraries/LibUnicode/Normalize.cpp

```diff
@@ -0,0 +1,196 @@
+                continue;
+            return mapping.code_point;
+        }
+    }
+    return 0;
+}
+
+enum class UseCompatibility {
+    Yes,
+    No
+};
+
+static void decompose_code_point(u32 code_point, Vector<u32>& output, UseCompatibility use_compatibility)
+{
+    // FIXME: Hangul characters are not handled.
+    auto const mapping = Unicode::code_point_decomposition(code_point);
+    if (mapping.has_value()) {
+        if (mapping->tag == CompatibilityFormattingTag::Canonical) {
+            decompose_code_point(mapping->decomposition[0], output, use_compatibility);
+            // "When a canonical mapping consists of a pair of characters,
+            //  the first character may itself be a character with a decomposition mapping,
+            //  but the second character never has a decomposition mapping."
+            if (mapping->decomposition.size() > 1 && mapping->decomposition[1] != 0) {
+                output.append(mapping->decomposition[1]);
+            }
+            return;
+        }
+        if (use_compatibility == UseCompatibility::Yes) {
+            for (auto code_point : mapping->decomposition) {
+                decompose_code_point(code_point, output, use_compatibility);
+            }
+            return;
+        }
+    }
+    output.append(code_point);
+}
+
+// This can be any sorting algorithm that maintains order (like std::stable_sort),
+// however bubble sort is the easier to implement, so go with it.
+template<typename T, typename LessThan>
+void bubble_sort(Span<T> span, LessThan less_than)
+{
+    for (size_t i = 0; i < span.size() - 1; ++i) {
+        for (size_t j = 0; j < span.size() - 1 - i; ++j) {
+            if (!less_than(span[j], span[j + 1]))
+                swap(span[j], span[j + 1]);
+        }
+    }
+}
+
+// The Canonical Ordering Algorithm, as specified in Version 15.0.0 of the Unicode Standard.
+// See Section 3.11, D109; and UAX #15 https://unicode.org/reports/tr15
+static void canonical_ordering_algorithm(Vector<u32>& input)
+{
+    for (size_t i = 0; i < input.size(); ++i) {
+        if (!is_starter(input[i])) {
+            auto starter = find_if(input.begin() + i, input.end(), is_starter);
+            auto const span_size = static_cast<size_t>(starter - (input.begin() + i));
+            // Nothing to reorder, so continue.
+            if (span_size <= 1)
+                continue;
+            Span<u32> span { input.data() + i, span_size };
+
+            bubble_sort(span, [](u32 a, u32 b) {
+                // Use <= to keep ordering.
+                return Unicode::canonical_combining_class(a) <= Unicode::canonical_combining_class(b);
+            });
+
+            // Skip over span we just sorted.
+            i += span_size - 1;
+        }
+    }
+}
+
+// See Section 3.11, D115 of Version 15.0.0 of the Unicode Standard.
+static bool is_blocked(Span<u32> span, size_t a, size_t c)
```

Comment:
`span` isn't a good variable name, let's call it `code_points` or something similar